src/Gemstone.Threading/NamedSemaphoreUnix.cs

```diff
@@ -210,23 +209,19 @@ public int ReleaseCore(int releaseCount)
         if (releaseCount < 1)
             throw new ArgumentOutOfRangeException(nameof(releaseCount), "Non-negative number required.");
 
-        int? previousCount = null;
+        int retVal = GetSemaphoreCount(m_semaphore, out int previousCount);
 
-        for (int i = 0; i < releaseCount; i++)
-        {
-            int retVal = GetSemaphoreCount(m_semaphore, out int count);
-            
-            if (retVal == ErrorNo.EINVAL)
-                throw new InvalidOperationException("The named semaphore is invalid.");
-
-            if (retVal != 0)
-                throw new InvalidOperationException($"An unknown error occurred while getting current count for the named semaphore. Error code: {retVal}");
+        if (retVal == ErrorNo.EINVAL)
+            throw new InvalidOperationException("The named semaphore is invalid.");
 
-            if (count >= m_maximumCount)
-                throw new SemaphoreFullException("The semaphore count is already at the maximum value.");
+        if (retVal != 0)
+            throw new InvalidOperationException($"An unknown error occurred while getting current count for the named semaphore. Error code: {retVal}");
 
-            previousCount ??= count;
+        if (previousCount >= m_maximumCount)
+            throw new SemaphoreFullException("The semaphore count is already at the maximum value.");
```

Comment:
It looks like it would it be possible to call this with a release count of zero, which I guess should just get the current count.

```suggestion
        if (releaseCount > 0 && previousCount >= m_maximumCount)
            throw new SemaphoreFullException("The semaphore count is already at the maximum value.");
```

---

Or should we preemptively avoid releasing the semaphore if we already know that it's going to exceed the maximum count? It seems wrong to allow the error to have side-effects, but it's hard to say which would be better in this case. The error is likely to have serious implications for the stability of the app regardless.

```suggestion
        if (previousCount + releaseCount > m_maximumCount)
            throw new SemaphoreFullException("The maximum count for the semaphore would be exceeded.");
```

Reply:
Question: for first item, wouldn't zero be caught by this:
```C#
        if (releaseCount < 1)
            throw new ArgumentOutOfRangeException(nameof(releaseCount), "Non-negative number required.");
```

For second item, the thinking here was that this was a basic check to make sure the maximum count was not already exceeded, i.e., a fast release at C# level without another Linux call.

For the actual check for exceeded count per each release iteration, was leaning on Linux return value here (see `EOVERFLOW` handling below and Linux manual page for [sem_post](https://www.man7.org/linux/man-pages/man3/sem_post.3.html)):

```C#
            retVal = ReleaseSemaphore(m_semaphore);

            switch (retVal)
            {
                case 0:
                    continue;
                case ErrorNo.EOVERFLOW:
                    throw new SemaphoreFullException("The maximum count for the semaphore would be exceeded.");
                case ErrorNo.EINVAL:
                    throw new InvalidOperationException("The named semaphore is invalid.");
                default:
                    throw new InvalidOperationException($"An unknown error occurred while releasing the named semaphore. Error code: {retVal}");
            }
```

Reply:
I guess the root question is whether to allow a `Release(5)` to succeed, in advance of actual release, when you know if will already fail because you can only release 4.

Gets tricky because of cross process race conditions...

Reply:
You're right. It's hard to say what I was thinking at this point, but if I were to guess, my mind was focused on how the for loop changed and didn't even notice the validation above it.

Reply:
> I guess the root question is whether to allow a `Release(5)` to succeed, in advance of any actual releases, when you know it will already fail because you can only release 4.
> 
> However, this gets tricky because of cross process race conditions...

Also, fyi, my thoughts on the matter: The point of the semaphore is to release only as many times as you wait so exceeding the semaphore count suggests that some user released the semaphore on behalf of another user who didn't know they were doing that. Regardless of race conditions, this represents a logic error, and it's impossible to know what the "correct" behavior is in the general case. So I've decided I'm not really worried about it either way.